### PR TITLE
352 imagnet model load into dashboard and make sure dashboard works

### PR DESCRIPTION
**Pull request by laurasootes**
The ImageNet model should now work with all three methods (and should still work for the the MNIST model). I also included buttons to set the parameters of each model. 

Some comments:

- I am not 100% that the output is correct, mainly because the output for the bee looks different from what the result of the tutorial. If someone could verify this that would be great
- In the tutorial the ResNet model is loaded from Keras, while here I downloaded a pretrained onnx model from https://github.com/onnx/models/tree/main/vision/classification/resnet/model to have the dashboard more generalised and satisfy with the original design. It could be that the results are therefore somewhat different? 
- To get the labels I still use the Keras implementation of ResNet, because the onnx file does not save labels. This makes the dashboard less generalised for, but can be solved in the future if we implement some sort of label upload. (see existing issue). That would btw also solve ugly non-generalised things such as 
if len(predictions[0]) == 2:
                class_name = class_name_mnist
- To use the pretrained onnx model I had the prepare the image to fit the model (resize the image and make sure the color channel is in the correct position). This is also a non-ideal non-generalised implementation.
- Besides the bee image, I also uploaded an image of a goldfish to try out.
- On my laptop the Lime model takes forever and ever to load. Not sure why.
- It is (easily) possible to change the layout of the parameter buttons. For example sliders (within a specific range), or a dropdown with specific values to chose from, radio items and more; see https://dash.plotly.com/dash-core-components

### CODE REVIEW

**Review by geek-yang — CHANGES_REQUESTED** [review]
Very nice work! It is good to see the parameter configuration works now and this makes the dashboard very informative and tweakable! I have some comments regarding coding styles and also some hard-coded parts.

Besides, here are some thoughts related to the questions in the top posts and something nice to have for the future:
- The computing time and results are very sensitive to the parameters, and even the resnet models (e.g. version v1.7 v.s. v2.7). So don't worry about it 😜 .
- When tweaking some parameters of the methods that are not selected, it seems the images are refreshing as well. Is there an easy option to disable this? Maybe this can be solved when we implement an `update` button. We can address it in another PR.
- Something nice to have for the future. We can implement collapse menu for the parameters (https://dash-bootstrap-components.opensource.faculty.ai/docs/components/collapse/) and hover-over explanations for all hyperparameters.

File: dashboard/layouts.py [hunk 1]
```diff
@@ -244,26 +244,190 @@ def get_uploads_images():
...
+                        placeholder = "Number of masks",
+                        type="number",
+                        value=1000,
+                        style={
+                                'margin-right': '20px',
+                                'margin-top': '0px',
+                                'width': '100%',
+                                'color' : colors['blue1']
+                            }
+                    ),
```
[the remark this answers is not in the document]
Comment (geek-yang) [reply dashboard/layouts.py:320]: image

Something like this?

**Comment by geek-yang** [comment]
More comments on the questions listed in the top post:

> * To use the pretrained onnx model I had the prepare the image to fit the model (resize the image and make sure the color channel is in the correct position). This is also a non-ideal non-generalised implementation.

This is ok. In practice, we expect the user to prepare their model and input images. Since the model is generated by them, they could decide where they would like to put the color channel. In our case, we simply grab some models from somewhere and therefore we had this issue.

> * On my laptop the Lime model takes forever and ever to load. Not sure why.

On my machine it works well. I'm using WSL. I guess this is somehow related to your environment.

File: dashboard/callbacks.py [hunk 2]
```diff
@@ -234,76 +245,83 @@ def update_multi_options_i(fn_m, fn_i, sel_methods, new_model, new_image):
...
                 class_name = class_name_mnist
-
-            pred_class = class_name[np.argmax(predictions)]
-
-            n_rows = len(class_name)
-
-            fig = make_subplots(rows=n_rows, cols=3, subplot_titles=("RISE", "KernelShap", "LIME"))  # , horizontal_spacing = 0.05)
-
+            else:
+                class_name = class_names_imagenet
```
[the remark this answers is not in the document]
Comment (laurasootes) [reply dashboard/callbacks.py:251]: yes, this is exactly what I was referring to in point three above ;)

File: dashboard/utilities.py [hunk 3]
```diff
@@ -100,9 +115,36 @@ def open_image(path):
...
+    img_data = keras_utils.img_to_array(img)
+    img_data = preprocess_input(img_data)
+    if img_data.shape[0] != 3:
+        # Colour channel is not in position 0; reshape the data
+        xarray = to_xarray(img_data, {0: 'height', 1: 'width', 2: 'channels'}) 
+        reshaped_data = move_axis(xarray, 'channels', 0)
+        img_data = np.array(reshaped_data)
+    # definitions for normalisation (for ImageNet)
+    mean_vec = np.array([0.485, 0.456, 0.406])
+    stddev_vec = np.array([0.229, 0.224, 0.225])
```
[the remark this answers is not in the document]
Comment (laurasootes) [reply dashboard/utilities.py:140]: source added in comments (https://github.com/onnx/models/tree/main/vision/classification/resnet)

File: dashboard/layouts.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (laurasootes) [reply dashboard/layouts.py]: created issue #412 to solve this

File: dashboard/callbacks.py [hunk 4]
```diff
@@ -1,29 +1,33 @@
...
+import onnx
+import plotly.express as px
 # Plotly
 import plotly.graph_objects as go
-from plotly.subplots import make_subplots
-import plotly.express as px
-# Dash&Flask
-from jupyter_dash import JupyterDash
-import dash
+import spacy
```
[the remark this answers is not in the document]
Comment (laurasootes) [reply dashboard/callbacks.py:12]: I added spacy, but for some reason it now fails the listing check :/

**Review by geek-yang — APPROVED** [review]
Thanks for addressing my comments. I think it looks good now.

One last comment before merging, could you please verify that the dashboard can be launched without trouble? I got some issues trying to launch the dashboard today. There are errors related dependency. Given that I just update my conda environment entirely, this could be my environment. If it runs without any problem, just feel free to press the button. Thanks for the effort! 😄

**Comment by laurasootes** [comment]
I checked running the dashboard, and it seems to launch without problems
